thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java

```diff
@@ -0,0 +1,320 @@
+            } catch (Exception e) {
+              LOG.error("Failed to run sla check on metric URN %s", datasetSLA.getKey(), e);
+            }
+          }
+        }
+      }
+
+      //LOG.info("End data availability for config {} between {} and {}. Detected {} anomalies.", config.getId(),
+      //    info.getStart(), info.getEnd(), anomaliesList.size());
+      return Collections.emptyList();
+    } catch(Exception e) {
+      throw e;
+    }
+  }
+
+  /**
+   * Runs the data sla check for the window (info) on the given metric (me) using the configured sla properties
+   */
+  private void runSLACheck(MetricEntity me, DatasetConfigDTO datasetConfig, DetectionPipelineTaskInfo info,
+      Map<String, Object> slaProps) {
+    if (me == null || datasetConfig == null || info == null) {
+      //nothing to check
+      return;
+    }
+
+    try {
+      long datasetLastRefreshTime = datasetConfig.getLastRefreshTime();
+      if (datasetLastRefreshTime <= 0) {
+        // assume we have processed data till the current detection start
+        datasetLastRefreshTime = info.getStart() - 1;
+      }
+
+      if (isMissingData(datasetLastRefreshTime, info)) {
+        // Double check with data source as 2 things are possible.
+        // 1. This dataset/source may not support availability events
+        // 2. The data availability event pipeline has some issue.
+        // We want to measure the overall dataset availability (filters are not taken into account)
+        MetricSlice metricSlice = MetricSlice.from(me.getId(), info.getStart(), info.getEnd(), ArrayListMultimap.<String, String>create());
+        DataFrame dataFrame = this.provider.fetchTimeseries(Collections.singleton(metricSlice)).get(metricSlice);
+        if (dataFrame == null || dataFrame.isEmpty()) {
+          // no data
+          if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+            createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+          }
+        } else {
+          datasetLastRefreshTime = dataFrame.getDoubles("timestamp").max().longValue();
+          if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+            if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+              createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+            }
+          }
+        }
+      } else if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+        // Optimize for the common case - the common case is that the data availability events are arriving
+        // correctly and we need not re-fetch the data to double check.
+        if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+          createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
```

Comment:
What we have so complicated logic here? Let's find some time to discuss offline.

Reply:
discussed offline